cpp/cycles_module/algorithm/cycles.cpp

```diff
@@ -0,0 +1,208 @@
+#include <unordered_set>
+
+#include <mg_exceptions.hpp>
+#include <mg_graph.hpp>
+
+#include "cycles.hpp"
+
+namespace cycles_util {
+
+NodeState::NodeState(std::size_t number_of_nodes) {
+  visited.resize(number_of_nodes, false);
+  parent.resize(number_of_nodes, std::nullopt);
+  depth.resize(number_of_nodes, 0);
+}
+
+void NodeState::SetVisited(std::uint64_t node_id) { visited[node_id] = true; }
+
+bool NodeState::IsVisited(std::uint64_t node_id) const { return visited[node_id]; }
+
+void NodeState::SetParent(std::uint64_t parent_id, std::uint64_t node_id) { parent[parent_id] = node_id; }
+
+std::optional<std::uint64_t> NodeState::GetParent(std::uint64_t node_id) const { return parent[node_id]; }
+
+void NodeState::SetDepth(std::uint64_t node_id, std::uint64_t node_depth) { depth[node_id] = node_depth; }
+
+std::uint64_t NodeState::GetDepth(std::uint64_t node_id) const { return depth[node_id]; }
+
+void FindNonSpanningTreeEdges(std::uint64_t node_id, const mg_graph::GraphView<> &graph, NodeState *state,
+                              std::set<std::pair<uint64_t, uint64_t>> *non_st_edges) {
+  std::unordered_set<std::uint64_t> unique_neighbour;
+  state->SetVisited(node_id);
+  for (const auto &neigh : graph.Neighbours(node_id)) {
+    auto next_id = neigh.node_id;
+
+    // Check if is returning edge or already visited neighbour
+    if (state->GetParent(node_id) != std::nullopt &&
+        (next_id == state->GetParent(node_id) || unique_neighbour.find(next_id) != unique_neighbour.end())) {
```

Comment:
`std::optional` has an implicit conversion to bool so you don't have to compare it to `!= std::nullopt` in ifs.
```suggestion
    if (const auto parent = state->GetParent(node_id); 
        parent && (next_id == parent || unique_neighbour.find(next_id) != unique_neighbour.end()) {
```